node/sync/src/block_sync.rs

```diff
@@ -381,14 +381,32 @@ impl<N: Network> BlockSync<N> {
 
         // Compute the common ancestor with this node.
         let mut ancestor = 0;
-        for (height, hash) in locators.clone().into_iter() {
-            if let Ok(canon_hash) = self.canon.get_block_hash(height) {
-                match canon_hash == hash {
-                    true => ancestor = height,
-                    false => break, // fork
+
+        // The unwrap is safe, as `ensure_is_valid` guarantees that `recents` is not empty.
+        let (latest_locator_height, latest_locator_hash) = locators.recents.last().unwrap();
+
+        // Check if the latest locator matches the canon to quickly determine if it's a fork.
+        // This is an optimization that allows us to skip the traversal of all the locators in
+        // case it's a match.
+        if let Ok(canon_hash) = self.canon.get_block_hash(*latest_locator_height) {
+            if *latest_locator_hash == canon_hash {
+                ancestor = *latest_locator_height;
+            }
+        }
```

Comment:
What happens if a malicious node sends garbage locators except for the most recent height? Then we will skip fork detection and won't discover this case.

Reply:
I don't know what the goal of such a node would be, but I may be missing the bigger picture. Please feel free to close this PR if this is a viable attack vector or if it invalidates some of the assumptions about our peers; we can always revisit this if block locator spam/DoS proves to be a greater concern.